### ISSUE TITLE
`Suppliers.memoize()` thread pinning

### ISSUE THREAD

**Issue by steven-sheehy** [google/guava#7140]
### API(s)

```java
`Suppliers.memoize()`
```

### How do you want it to be improved?

Change it from `synchronized` to `ReentrantLock`

### Why do we need it to be improved?

`Suppliers.memoize()` internally uses the double-checked locking idiom with the `synchronized` keyword. Using `synchronized` with virtual threads can cause thread pinning and slow down performance.

### Example

```java
`Suppliers.memoize(() -> someLongRunningTask());`
```

### Current Behavior

Functions properly but causes performance degradation when used with virtual threads on Java 21

### Desired Behavior

Performs optimally with virtual threads.

### Concrete Use Cases

In our virtual thread enabled REST API, added lazy loading cache using `Suppliers.memoize()` around some of our database interactions to improve performance but ended up seeing worse performance than going directly to database.

### Checklist

- [X] I agree to follow the [code of conduct](https://github.com/google/.github/blob/master/CODE_OF_CONDUCT.md).

- [X] I have read and understood the [contribution guidelines](https://github.com/google/guava/wiki/HowToContribute#feature-requests).

- [X] I have read and understood [Guava's philosophy](https://github.com/google/guava/wiki/PhilosophyExplained), and I strongly believe that this proposal aligns with it.

**Comment by ben-manes** [comment]
This is already addressed by the JDK team’s [EA build](https://mail.openjdk.org/pipermail/loom-dev/2024-February/006433.html). That’ll avoid surprises as a better long term solution.

**Comment by steven-sheehy** [comment]
That's cool to see and would definitely help long term, but unfortunately my company only uses LTS versions of Java so that'd mean we'd have to wait until Java 25 (or whatever it is) to be released.

**Comment by ben-manes** [comment]
If you read the thread, you’ll see that VTs still have [serious problems](https://mail.openjdk.org/pipermail/loom-dev/2024-February/006463.html) and must be used very carefully for now. You might not want to use them until 25 when hopefully they’ll be less error prone. Libraries trying to support them early might not really help; imho it’s best to wait.

> the summary here is that addressing the issue with object 
monitors pinning is great but the hoorays may be short lived as the spot 
light moves to other cases where carriers are pinned, and specifically 
native frames due to resolving references to classes in the constant 
pool and the resulting class loading, or class initializers. There are 
some ideas around this that may provide some relief on these cases. We 
had to shake out issues with object monitors first.

**Comment by cpovirk** [comment]
+1, thanks.

One other note: We did a little work to improve the memory usage of `Suppliers.memoize` recently. I would suspect that a switch to `ReentrantLock` would more than undo the gains from that. And the reason that we did the optimization is that it appeared to matter at least a little to our fleet in aggregate, so I'd expect a switch to `ReentrantLock` hurt in a small but measurable way.

There is probably a way to write our own _even more efficient_ implementation of `Suppliers.memoize`, which would use `LockSupport.park` and friends directly. (I'd played around with something similar for Dagger in my internal cl/209143332, as has another developer in cl/448147583.) That could not only be more compact but also avoid making waiters take turns in receiving the produced value. It hasn't been clearly worth the effort, but it's possible that virtual threads will change the calculus.

Whatever we decide for now, it's possible that we'll revisit as we get more experience with virtual threads. (If we encounter the level of problems that Ben Manes passes along, then we may not get that experience anytime soon :))

**Comment by cpovirk** [comment]
Oh, also: As the Java platform develops more broadly, we may also see alternatives to `Suppliers.memoize` become available there, reducing the need for Guava's version. See, for example, https://openjdk.org/jeps/8312611 or even https://openjdk.org/jeps/8209964.

**Comment by cpovirk** [comment]
@lowasser reports from JVMLS that, while the pinning fix will be in Java 25, `synchronized` will still be _slower_ than explicit lock objects. (It isn't clear whether the performance gap is expected to shrink or disappear over time.)

That is not necessarily highly relevant to users of `Suppliers.memoize`, though: The implementation of that method [skips the lock](https://github.com/google/guava/blob/5c23347cc162a1dcb046b034ba8db1db4f0de77b/guava/src/com/google/common/base/Suppliers.java#L143) after the initial setup, so the slowness might not be an issue. That said, it's always possible that the cost of that initial setup, if bad enough, could be a problem for startup times or other latency measurements, so we'd listen to stories of real-world impact. We'd still need to consider any increase in _RAM_ usage that results from additional lock objects.

(I had also gotten to wondering whether some _other_ JDK lock implementation might provide what we need without the footprint of `ReentrantLock`. But the other one that came to mind is `StampedLock`, which also looks relatively large.)

The story for _other_ Guava uses of `synchronized` could always be different. The one that comes to mind is the one in `cache`, but we had [some misadventures there last year](https://github.com/google/guava/issues/6845), and we stick by our recommendation to use [Caffeine](https://github.com/ben-manes/caffeine) instead.

**Comment by cpovirk** [comment]
(Oh, and I was going to throw in: `Suppliers.memoize` could perhaps someday use `StableValue` once it's finalized. But I notice that it (at least currently) [uses `synchronized`](https://github.com/openjdk/jdk/blob/72d1066ae385ce511a70e762ed68afaa72e96ec4/src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java#L87) :))

**Comment by cpovirk** [comment]
(Perhaps accordingly, I am seeing a `StableValue`-based implementation of `Suppliers.memoize` perform worse than the current implementation _in a create-and-compute benchmark_. It appears similarly performant when getting an already completed value (though I could have dreamed that we'd see better).)

**Comment by cpovirk** [comment]
https://github.com/netty/netty/issues/16158 sounds maybe a little like `synchronized` is still not ideal for virtual threads in some cases? I haven't tried to understand deeply, but it's something for us to continue to keep an eye on. I wouldn't mind an excuse to dust off cl/209143332 and rework it for `Suppliers`, but so far, I think we're lacking in evidence that that would be a productive use of time :)

**Comment by MarcoLe3** [comment]
Hi, I thought this discussion was pretty interesting and dove a little deep into the topic to understand it more. I don't have a full grasp of guava yet but I was wondering on your opinion if `Supplier.memoize` used CAS instead for virtual threads and have `Locksupport.park() `and `unpark()` for other threads that are waiting? 

Unless this was ruled out in a different discussion.

**Comment by cpovirk** [comment]
Yep, manual parking [would be cool](https://github.com/google/guava/issues/7140#issuecomment-2045829722). At this point, I don't think it's justified, as fun as it would be: I don't want to risk having someone important come to me during a postmortem and having to explain that, yeah, we didn't actually think the performance gain was significant, but we wanted to try it :)

**Comment by MarcoLe3** [comment]
Ah, I see haha. I wonder what kind of other techniques that would provide a significant boost in performance when dealing with virtual threads.
